core/store/src/lib.rs

```diff
@@ -404,6 +407,29 @@ impl StoreUpdate {
         }
         self.storage.write(self.transaction)
     }
+
+    pub fn apply_change_to_flat_state(&mut self, change: &RawStateChangesWithTrieKey) {
+        match &change.trie_key {
+            TrieKey::DelayedReceipt { .. } | TrieKey::DelayedReceiptIndices => return,
+            _ => {}
+        }
+        let key = change.trie_key.to_vec();
+        let last_change = change
+            .changes
+            .last()
+            .expect("Committed entry should have at least one change")
+            .data
+            .clone();
+        match last_change {
+            Some(value) => {
+                let mut value_ser = [0u8; 36];
+                value_ser[0..4].copy_from_slice(&(value.len() as u32).to_le_bytes());
+                value_ser[4..36].copy_from_slice(&hash(&value).0);
+                self.set(DBCol::FlatState, &key, &value_ser)
```

Comment:
Should we also think about storing values directly in flat state if values are short?

```suggestion
                let mut buffer = [0u8; 128];
                let (len_buf, value_buf) = buffer.split_at_mut(4);
                len_buf.copy_from_slice(&(value.len() as u32).to_le_bytes());
                let len = 4 + if value.len() <= value_buf.len() {
                    value_buf[..value.len()].copy_from_slice(&value);
                    value.len();
                } else {
                    value_buf[..32].copy_from_slice(hash(&value).as_ref());
                    32
                };
                self.set(DBCol::FlatState, &key, &buffer[..len])
```

Though perhaps that’s not worth considering at this early stage.

Reply:
Good point. I would consider it after shipping flat storage, because we can switch to storing values using quite simple migration.